### PR TITLE
 [Constraint solver] Look through optional binding for overload sets.

### PR DESCRIPTION
**Pull request by DougGregor**
When we’re trying to find the overload set corresponding to a particular
type variable, look through “optional object of” constraints that represent
the use of ? binding or ! forcing. This allows us to find overload sets
when referring to, e.g., @objc optional protocol requirements.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: lib/Sema/CSSolver.cpp [hunk 1]
```diff
@@ -1907,9 +1907,73 @@ void ConstraintSystem::partitionForDesignatedTypes(
...
+
+    auto funcType = AFD->getInterfaceType();
+    auto hasAnyOrOptional = funcType.findIf([](Type type) -> bool {
+      if (type->getOptionalObjectType())
+        return true;
+
+      return type->isAny();
+    });
+
+    // If function declaration references `Any` or `Any?` type
```
Comment (slavapestov) [inline lib/Sema/CSSolver.cpp]: You're looking for `$x?` for any `$x`, not `Any?` here

File: lib/Sema/CSSolver.cpp [hunk 2]
```diff
@@ -1907,9 +1907,73 @@ void ConstraintSystem::partitionForDesignatedTypes(
...
+    // let's not attempt it, because it's unclear
+    // without solving which overload is going to be better.
```
Comment (slavapestov) [inline lib/Sema/CSSolver.cpp]: Is there a test case that fails without this hack?

**Review by slavapestov — APPROVED** [review]

**Comment by DougGregor** [comment]
There's a problem with my PR in https://github.com/apple/swift/pull/23114 that I need to address there, first; then I'll build on it with this.

File: lib/Sema/CSSolver.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/CSSolver.cpp]: I'll clean up the comment for a future PR.

File: lib/Sema/CSSolver.cpp (hunk 2 above)
Comment (DougGregor) [reply lib/Sema/CSSolver.cpp]: Yeah, this was https://bugs.swift.org/browse/SR-7884.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Pull request merged**